Blog Post - A Unified Theory of Garbage Collection (#119)
sampsyo/cs6120: content/blog/2019-11-08-unified-theory-gc/index.md

## Background

Broadly speaking, garbage collection (GC) is a form of automatic memory management. The garbage collector attempts to free the memory blocks occupied by objects that are no longer in use by the program. It relieves programmers from the burden of explicitly freeing allocated memory. Moreover, it also serves as part of the security strategy of languages like Java: in Java virtual machine programmers are unable to accidentally (or purposely) crash the machine by incorrectly freeing memory. The opposite is manual memory management, which is available in C/C++. This gives the maximum freedom for programmers and avoids the potential overhead that affects program performance.

The task garbage collection needs to solve is identifying the objects not accessible by the program in the reference graph. It then frees the unreachable objects and rearranges the memory sometimes to reduce heap fragmentation. 

The most traditional approaches are tracing and reference counting:

- Tracing: Recursively mark reachability by starting from a set of roots memory

>   blocks that are in use (e.g. pointed by global variable or local variable

Reviewer: ```suggestion
  blocks that are in use (e.g., pointed to by global variables or local variables
```
https://capra.cs.cornell.edu/styleguide/#egie

  currently in stack frames).
- Reference Counting: Count the number of pointers pointed to one particular
  object by bookkeeping it every time a pointer is created or modified. It frees
  the object when the counter decreases to zero.

These two approaches have a lot differences:

<img src="diff.png" style="width: 100%">

> Although tracing naturally solves the reachability problem accurately, it requires to traverse over a static graph and therefore suspend the whole program. On the other hand, reference counting is done incrementally along with each pointer assignment and collect. However, it brings unnecessary overhead when the pointers are changed often and it does not collect cycles of garbages. Thus people proposed more complicated algorithms based on different hypotheses, such as deferred reference counting, generational garbage collection, etc. 

Reviewer: ```suggestion
Although tracing naturally solves the reachability problem accurately, it requires to traverse over a static graph and therefore suspend the whole program. On the other hand, reference counting is done incrementally along with each pointer assignment and collect. However, it brings unnecessary overhead when the pointers are changed often and it does not collect cycles of garbage. Thus people proposed more complicated algorithms based on different hypotheses, such as deferred reference counting, generational garbage collection, etc. 
```

counting collector does not detect any cyclic garbage, and thus finds the
greatest fixed point.

### Alignments of Algorithmic Structures

As modeled in the paper, a tracing garbage collector fundamentally works by
performing a traversal on the memory graph, starting from the root objects in
$R$. Every time it reaches an object, it increments its reference count. And,
when it reaches an object for the first time, it traverses the pointers from
that object. In order to complete the traversal, the algorithm maintains a work
list, $W$, of objects that it needs to visit. This traversal recomputes all
reference counts completely each time it runs. When it terminates, every live
object will have a positive reference count, and every dead object will have a
reference count of zero.

This formulation of the algorithm differs in a few ways from a standard tracing
collector. For example, there is no need in a real tracing collector to keep
track of the full reference count of each object; it is sufficient to keep just
a single mark bit which determines whether the reference count is non-zero. A
tracing collector might also employ other optimizations, such as pointer

> reversal, to improve its run time or memory usage. However, these optimizations

Reviewer: Pointer reversal is sort of an advanced topic we haven't talked about here—maybe just link to a basic explanation somewhere?

represent its relationship to tracing garbage collection. Formulated in this
way, if we look at the code for the central work-list algorithms for each type
of garbage collector, we can see striking similarities between them:

<img src="tracing-wl.png" style="width:100%"> | <img src="rc-wl.png" style="width:100%"> 
:-------------------------------------:|:--------------------------:
 Graph traversal algorithm for tracing | Work-list algorithm for RC

The only differences between the two are that tracing increments the reference
counts of objects on the worklist whereas RC decrements them, and that tracing
checks if $\rho(w) = 1$ for when to add the objects that $w$ points to whereas
RC checks if $\rho(w) = 0$. The two garbage collection strategies also differ in
how $W$ is initialized. For tracing, $W$ initially contains all of the roots of
the graph, $R$. For reference counting, $W$ starts with all of the objects that
have had a pointer to them erased since the last time the algorithm ran. By
viewing the two strategies through this formulation, we can see that they are
opposites of each other in many ways:

<img src="diff-v2.png" style="width:100%">

> As discussed in the mathematical model part, reference counting requires an extra pass to collect cycles. This is generally be done by two strategies: backup tracing, which traces (part of) the heap occasionally, and trial deletion, which attempts to decrease  $\rho(w)$ by trial-and-error guided by heuristics. However, notice that there is also counterpart for that in tracing: sweeping phase. In addition, while tracing converges to the fixed point value stating from the lower bound 0, reference counting starts from the upper bound, which is all incoming pointers that existed historically. 

Reviewer: ```suggestion
As discussed in the mathematical model part, reference counting requires an extra pass to collect cycles. This is generally be done by two strategies: backup tracing, which traces (part of) the heap occasionally, and trial deletion, which attempts to decrease  $\rho(w)$ by trial-and-error guided by heuristics. However, notice that there is also counterpart for that in tracing: the sweeping phase. In addition, while tracing converges to the fixed point value stating from the lower bound 0, reference counting starts from the upper bound, which is all incoming pointers that existed historically. 
```

Reviewer: Perhaps instead of "the lower bound 0," consider "the lower bound $\varnothing$", if you mean the empty set?

## Hybrids

The authors further show that all realistic garbage collectors are in fact hybrids of tracing and reference counting. In general, we can categorize collectors to unified heap collectors, split heap collectors, and multi-heap collectors. Then different garbage collectors can be seen as performing tracing or reference counting when tracking references within each region and across regions.

### Unified Heap collectors: Deferred Reference Counting & Partial Tracing

<img src="deferred.png" alt="Snow" style="width:100%"> |  <img src="partial.png" alt="Snow" style="width:100%">
:-------------------------:|:-------------------------:
Deferred Reference Counting |  Partial Tracing

> Rather than doing reference counting completely, Deferred Reference Counting defers updating the reference count of objects pointed directly by roots until batch processing. This is based on the observation that pointers from roots are likely to change very often as they are directly used in the program. Notice that we can view this as tracing from roots to their targets and reference counting for the intra-heap pointers: All the assignments that lead to intra-heap pointers change would be tracked by reference counting as normal. When we suspend the program, we trace the roots for one level, which compensates for the delay.

Reviewer: ```suggestion
Rather than doing reference counting completely, Deferred Reference Counting defers updating the reference count of objects pointed directly by roots until batch processing. This is based on the observation that pointers from roots are likely to change very often as they are directly used in the program. Notice that we can view this as tracing from roots to their targets and reference counting for the intra-heap pointers: All the assignments that lead to intra-heap pointer changes would be tracked by reference counting as normal. When we suspend the program, we trace the roots for one level, which compensates for the delay.
```

> Reversely, we could design Partial Tracing, which uses reference counting for edges from roots to heaps while tracing the intr-heap pointers. However, this combines the worse properties of both tracing and reference counting: it suffers from the high mutation cost from the fast-changing of root pointers while still need to spend a long time to trace the heap. This design failure demonstrates that although tracing and reference counting are duals, they are not equally easy to solve under different cases. 

Reviewer: ```suggestion
Reversely, we could design Partial Tracing, which uses reference counting for edges from roots to heaps while tracing the intra-heap pointers. However, this combines the worst properties of both tracing and reference counting: it suffers from the high mutation cost from the fast-changing of root pointers while still need to spend a long time to trace the heap. This design failure demonstrates that although tracing and reference counting are duals, they are not equally easy to solve under different cases. 
```

> ### Split Heap and multi-Heap Collectors

Reviewer: ```suggestion
### Split Heap and Multi-Heap Collectors
```

> The most common Split-heap collector architecture is Generational collectors. Generational collectors are based on the following empirical observation that most objects are short lived, as shown in the left figure bellow. Here Y axis shows the number of bytes allocated and the X access shows the number of bytes allocated over time. ()

Reviewer: No need to capitalize "Split-heap" or "Generational collectors." (Please apply to the text below as well—things like "Multi-heap" also appear.)

> The most common Split-heap collector architecture is Generational collectors. Generational collectors are based on the following empirical observation that most objects are short lived, as shown in the left figure bellow. Here Y axis shows the number of bytes allocated and the X access shows the number of bytes allocated over time. ()

Reviewer: ```suggestion
The most common Split-heap collector architecture is Generational collectors. Generational collectors are based on the empirical observation that most objects are short lived, as shown in the left figure bellow. Here the Y axis shows the number of bytes allocated and the X access shows the number of bytes allocated over time. ()
```

> The most common Split-heap collector architecture is Generational collectors. Generational collectors are based on the following empirical observation that most objects are short lived, as shown in the left figure bellow. Here Y axis shows the number of bytes allocated and the X access shows the number of bytes allocated over time. ()

Reviewer: Looks like a stray () at the end of this paragraph?

This process can also be seen as a combination of tracing and reference counting as shown by (a) in the right figure: Reference counting is performed to track edges from mature space to nursery space. Tracing is performed within nursery space during minor collections. And finally a full tracing is performed during major collections.

We can then explore different combinations of tracing and reference counting within each space. However, notice that reference counting is always used for the edges from mature to nursery space in order to avoid visiting all objects in the mature space. In fact, the authors claim that any algorithm
that collects some subset of objects independently is fundamentally making use of reference counting. 

> This conclusion is further used for Multi-heap collectors, which split heaps into multiple regions and collect each region individually. They can be treated asymmetrically (as in Generational collector) or symmetrically. One of the basic extension from Generational collector is the Train algorithm. The train algorithm extended the mature space to multiple regions of different "generations", where only one region is collected at a time. This is mainly for reducing the pause time for collecting the mature space. Specifically, it divides the Mature Object Space(MOS) to cars of fixed size, which are chained to form trains. Each time, the current youngest car will be collected, until the whole first train is collected. We assume that then this process moves on to the next train, but then it is unclear to us why we need several trains instead of one train with all cars?

Reviewer: ```suggestion
This conclusion is further used for Multi-heap collectors, which split heaps into multiple regions and collect each region individually. They can be treated asymmetrically (as in Generational collector) or symmetrically. One of the basic extension from Generational collector is the Train algorithm. The train algorithm extended the mature space to multiple regions of different "generations", where only one region is collected at a time. This is mainly for reducing the pause time for collecting the mature space. Specifically, it divides the Mature Object Space (MOS) to cars of fixed size, which are chained to form trains. Each time, the current youngest car will be collected, until the whole first train is collected. We assume that then this process moves on to the next train, but then it is unclear to us why we need several trains instead of one train with all cars?
```

discussed in the paper, they give equations or approximations of $\sigma$. As an
example, for a tracing collector, $\sigma(T) \simeq \frac{\mathcal{M}}{\omega} +
\mathcal{D}$, where $\mathcal{M}$ is the capacity of memory in units of the
object size, $\omega$ is the size of an object in bits, and $\mathcal{D}$ is the
space cost of the traversal stack, proportional to the traversal depth. This
model represents a version of tracing collection without pointer reversal, and
hence includes the traversal stack. If pointer reversal were considered,
$\mathcal{D}$ could be removed from the space equation, but this would also
increase the time cost of the collector due to the additional graph traversal
required.

They define $\tau(X)$ to be the time overhead of a garbage collector. For each
collector, they define $\tau$ in terms of a linear function of various
properties of the program, omitting the constant factors that could vary between
implementations. In general, they define the overhead as $\tau(X) =
\phi(X)\kappa(X) + \mu(X)$, where $\phi$ is the frequency at which collection
occurs, $\kappa$ is the run time of a single garbage collection execution, and

> $\mu$ is the overhead for mutation in the program (e.g. incementing and

Reviewer: ```suggestion
$\mu$ is the overhead for mutation in the program (e.g., incrementing and
```

reference counting collectors, we can consider the design of a high-performance
garbage collector as a balance between the two base strategies. And, by modeling
the time and space costs of different collectors, we can formally analyze the
trade-offs between different collection strategies for a specific workload or in
general. The paper suggests that the design of a garbage collector involves
three major decisions: how to partition memory, how to traverse memory, and what
space-time trade-offs should be used. The heap can be divided into multiple
segments, as with generational and multi-heap collectors, or can be collected
uniformly. Each component of the heap, as well as the roots of the graph, can
then be traversed and collected through either tracing or reference counting.
And within the implementation of the two strategies there are other space-time
trade-offs to consider as well, such as whether to employ pointer reversal.

In general, this paper provides a beautiful perspective that unifies garbage collectors to a spectrum between tracing and reference counting. It affects how garbage collectors are viewed afterward and the development of other unified theories in systems. 

> However, there are several aspects ignored in this paper too. Note that in this paper, the authors are only concerned with identifying unreachable objects correctly with high performance in terms of speed and space usage. This assumes that malloc can always do a perfect job to avoid memory fragmentation. But some garbage collectors are also designed to compact objects, such as Copying garbage collector.

Reviewer: ```suggestion
However, there are several aspects ignored in this paper too. Note that in this paper, the authors are only concerned with identifying unreachable objects correctly with high performance in terms of speed and space usage. This assumes that malloc can always do a perfect job to avoid memory fragmentation. But some garbage collectors are also designed to compact objects, such as copying garbage collectors.
```

It's not exactly "such as" copying collectors—that is the very definition of copying collectors. :smiley:

In addition, this formulation also ignores reachable memory leaks fundamentally. Although the memory manager can recover unreachable memory, it cannot free memory that is still reachable and therefore potentially still useful. Modern memory managers therefore provide techniques for programmers to semantically mark memory with varying levels of usefulness, which correspond to varying levels of reachability.

> Finally, there are also other methods that does automatic memory management partially in/before compile time: [abstract garbage collection](http://drops.dagstuhl.de/opus/volltexte/2019/10802/pdf/LIPIcs-ECOOP-2019-10.pdf), which soundly over-approximates the behaviour of a concrete

Reviewer: "other methods that *do*" (plurality agreement)

> interpreter) and [self-collecting mutators](https://dl.acm.org/citation.cfm?id=1993493) , which imposes extra invariants to help garbage collection.

Reviewer: Stray closing parenthesis?
```suggestion
interpreter and [self-collecting mutators](https://dl.acm.org/citation.cfm?id=1993493), which imposes extra invariants to help garbage collection.
```
